### DISCUSSION TITLE
Support for Functional Programming Paradigms

### DISCUSSION THREAD

**Discussion by Mastcharub** [carbon-language/carbon-lang#1687] [Language design]
I think that it might be a really good and nice idea to implement some of the Functional Programming Paradigms to Carbon such as: **Pure Functions**, **First Class/Higher Order Functions** and etc. to add a bit of more variety of possible patterns.
Hoping that other fellow Functional Programming Devs accepts this idea.
Edit: naturally the paradigms **must** be optional so that oop devs are not offended ;)

**Comment by junlarsen** [comment]
There has been a lot of discussion about lambda functions, semantics and syntax on the Discord lately, would definitely check it out if you're interested.

**Comment by MichaelBonnet** [comment]
I'd wager that such an effort would be much further down the road than where we are now, which is still in syntax specification type stuff. I can definitely see it happening, though.

**Comment by YagaoDirac** [comment]
In fact, c++ has a lot feature to handle functional programming. But the functor pattern is not famous at all. It's only because the feature is not famous. I used the pattern in one of my personal project and it was really good. But I've almost never seen it anywhere else.
Lambda is only about how to write a function at a specific position inside the code. In c++, without lambda, you can also do in-function-struct and define member function for that struct, or simple try functor.
C++ provides features for as much possibilities and let people to implement whatever they prefer. 
Before c++20, it's still pretty easy to make generators, and also the equivalant behavior of pipe operator(since c++20). So, your proposal maybe fit some kinda standard library better than built-in language feature.
